### ISSUE TITLE
bad math expression within Git repo

### ISSUE THREAD

**Issue by davidtwco** [sindresorhus/pure#373]
### General information

- Pure version: 1.6.0
- ZSH version: 5.4.2
- Git version: 2.15.0
- Terminal program & version: Windows Console (Bash for Windows)
- Operating system: Windows 10
- ZSH framework: Antibody

I have:
- [ ] Tested with another terminal program and can reproduce the issue: N/A
- [x] Followed the [Integration](https://github.com/sindresorhus/pure#integration) instructions for my framework

### Problem description
When in a folder with a git repo (I think it's only with git repos), I get the following error:

```
~/projects/somerepo
❯
prompt_pure_async_callback:79: bad math expression: illegal character: \
```

### Reproduction steps

1. Use same versions as above.
2. Go to a repo.
3. Run commands half a dozen or so times until it shows up.

### My `.zshrc`:
```shell
autoload -U promptinit; promptinit
prompt pure
```

**Comment by mafredri** [comment]
Are you sure you're running `v1.6.0`? The `prompt_pure_async_callback` function only has 76 lines in that version, but this error is coming from 79. If you're running `master`, please make sure `zsh-async` is up-to-date as well.

**Comment by mafredri** [comment]
I just tried Pure on Windows 10 with Bash for Windows (just installed), git version 2.7.4 and I'm unable to reproduce the issue. Tried with clean zsh (basically your `.zshrc` as posted) and using [my own zsh config](https://github.com/mafredri/zeet).

**Comment by davidtwco** [comment]
Thanks for looking into it, I'm pretty sure I'm running the latest `master` - I just set this machine up today and antibody downloaded both this and `zsh-async` from GitHub. I only said `1.6.0` as I looked at the `package.json` and that was the version listed.

After using this machine for most of the day, I've not ran into the issue after the handful of times initially so I'm not quite sure what was causing it. I'll close this issue for now and re-open if I run into it again and can do some looking into it.

Thanks again.

**Comment by davidtwco** [comment]
Ran into this again. I think it's only happening on certain repositories though.

I think I've figured out what the issue is though. This particular repository uses Git over SSH, the host it connects to is defined in the ssh config, it uses the OpenSSH 7.3 `ProxyJump` directive to connect to a host that can reach where the repo is stored. That host, for reasons outside of my control, only has a interactive password prompt - it can't use an SSH key.

Can you think of any ways that Pure might be able to detect this and not run a check on that repository, @mafredri?

**Comment by mafredri** [comment]
@davidtwco glad you found the source of this! I don't have any ideas for detecting this in Pure at the moment. I would need to dive deeper into what's possible with Git / SSH. Not sure when I'll get around to it, at least not for a week.

While not optimal, you could use a `vcs_info` configuration trick to disable git checks entirely, as described here: https://github.com/sindresorhus/pure/issues/189#issuecomment-173844254. This would disable branch and dirty checks as well though.

**Comment by davidtwco** [comment]
In the unlikely event anyone else is having this issue, I've managed to workaround it by adding the following to `~/.ssh/config` so that any sessions are re-used. That way I can keep a terminal open with a connection and while I do, the problems disappear.

```
ControlMaster auto
ControlPath /tmp/ssh_mux_%h_%p_%r
```

**Comment by eoli3n** [comment]
Same issue here, ControlMaster fix doesn't work for me. It appears only with ssh.
I also tried to add in my .zshrc
```
export ZSH_COMPDUMP=/tmp/zcompdump-$USER
source $ZSH/oh-my-zsh.sh
```
Without more success

**Comment by mafredri** [comment]
@eoli3n do you use some form of pinentry or other authentication prompt for your ssh repositories? If so, https://github.com/sindresorhus/pure/issues/366 could be related.

**Comment by eoli3n** [comment]
I do, when i push my git repo, yes

**Comment by eoli3n** [comment]
up

**Comment by mafredri** [comment]
@eoli3n the only workarounds I know of are the ones described in this issue, https://github.com/sindresorhus/pure/issues/373#issuecomment-348710294 and https://github.com/sindresorhus/pure/issues/373#issuecomment-349282099. I don't have a setup to test or experiment with this, so it's a pretty low priority issue for me at the moment.

I'm hoping someone else can chip in.

**Comment by eoli3n** [comment]
[![asciicast](https://asciinema.org/a/snEN2xTF7W64gN4SeXhpwC14b.png)](https://asciinema.org/a/snEN2xTF7W64gN4SeXhpwC14b)

https://github.com/sindresorhus/pure/issues/373#issuecomment-348710294 -> "This would disable branch and dirty checks as well though." : i need branch check
https://github.com/sindresorhus/pure/issues/373#issuecomment-349282099 -> not working

It's working on https://github.com/sindresorhus/pure/commit/f23e72640f02db0817cdb6d5d982b2c8f2f34656
[![asciicast](https://asciinema.org/a/jAR21PRxvNIhAUlo7fbTUZaW8.png)](https://asciinema.org/a/jAR21PRxvNIhAUlo7fbTUZaW8)

**Comment by mafredri** [comment]
In the second video, it's working because the commit you checked out does not have an upstream branch. That means there's no `git fetch` going on.

As mentioned in https://github.com/sindresorhus/pure/issues/366#issuecomment-348634030, the reason you're seeing the error is that an interactive program (e.g. pinentry) starts up (during `git fetch`) inside the async worker, essentially breaking it. The solution is to prevent an interactive program from starting up.

**Comment by eoli3n** [comment]
Sorry i don't get what you mean...
The only difference between 2 videos is the commit checked out in pure project.

Do you mean that last pure master commit does git checks not the same way than previous commit ?

Project ``/etc/ansible/playbooks-awx`` is hosted on a gogs srv. Does git checks hangs on authentication ?
I tried to export ``GIT_TERMINAL_PROMPT=0`` in zshrc with master pure commit, not working better...

``Pinentry`` seems related to ``gnupg``, i don't use it. I use ``ssh-agent``.

Anyway, i authenticate on my gogs srv with web authentication.
```
/etc/ansible/playbooks-awx master root@prodpeda-cesium
❯ git remote show origin 
Username for 'https://gogs.domain.fr': user.name@domain.fr
Password for 'https://user.name@domain.fr@gogs.domain.fr': 
* distante origin
  URL de rapatriement : https://gogs.domain.fr/gogs/p00000012366/playbooks-awx.git
  URL push : git@localhost:playbooks-awx.git
  URL push : https://domain.fr/gogs/p00000012366/playbooks-awx.git
  Branche HEAD : master
  Branches distantes :
    dev    suivi
    master suivi
  Branche locale configurée pour 'git pull' :
    master fusionne avec la distante master
  Références locales configurées pour 'git push' :
    dev    pousse vers dev    (à jour)
    master pousse vers master (à jour)
```

**Comment by mafredri** [comment]
It's probably a happy accident that the old version of Pure "works" (i.e. doesn't spit out errors). Can you confirm that you have an up-to-date zsh-async loaded in your shell?

Also, does `GIT_TERMINAL_PROMPT=0 git remote show origin` prompt for login? If it does, then that's the problem.

**Comment by eoli3n** [comment]
```
/etc/ansible/playbooks-awx master root@prodpeda-cesium
❯ GIT_TERMINAL_PROMPT=0 git remote show origin
Username for 'https://gogs.domain.fr': ^C
```

**Comment by mafredri** [comment]
Yeah, the fact that it prompts for input is the problem (messes with the async worker). Somehow that needs to be disabled, I don't know how.

I could silence the error in Pure, but that would only hide the problem.

Out of curiosity, is your git version older than `2.3`? If so, updating git would help.

**Comment by eoli3n** [comment]
It's on debian 8
```
❯ git --version                                   
git version 2.1.4
```
And it's a production srv so, no testing repo.

**Comment by eoli3n** [comment]
I think hiding can be better than actual behaviour

**Comment by mafredri** [comment]
I do understand your frustration, but hiding the problem will have repercussions. There's a chance that Pure's async functionality will stop working entirely if interactive programs are started inside the worker. I prefer seeing bug reports, specific like this one, over "pure suddenly stopped working".

At this moment I don't have a better solution for you than upgrade git or use `disable-patterns` for `vcs_info` (linked to earlier).

I wont consider hiding the error. Maybe we could show a nicer error message, but this is a very rare edge case.

**Comment by eoli3n** [comment]
I understand, but prompting the error as a new line seems not cool. A prompt should not be a log "file".
I will try a bit later, thx.

**Comment by robdodson** [comment]
Hitting this issue as well on GitHub repos cloned with https for what it's worth.

**Comment by robdodson** [comment]
```
❯ GIT_TERMINAL_PROMPT=0 git remote show origin
* remote origin
  Fetch URL: https://github.com/WICG/focus-visible.git
  Push  URL: https://github.com/WICG/focus-visible.git
  HEAD branch: master
  Remote branches:
    add-tests                                    tracked
    ae-fixup                                     tracked
    docs-cleanup                                 tracked
    fix-tests                                    tracked
    greenkeeper/chromedriver-2.35.0              tracked
    greenkeeper/eslint-4.16.0                    tracked
    greenkeeper/eslint-config-prettier-2.9.0     tracked
    greenkeeper/http-server-0.11.0               tracked
    greenkeeper/husky-0.15.0-rc.2                tracked
    greenkeeper/lint-staged-6.0.1                tracked
    greenkeeper/lint-staged-6.1.0                tracked
    greenkeeper/mocha-5.0.0                      tracked
    greenkeeper/prettier-1.10.0                  tracked
    greenkeeper/prettier-1.10.1                  tracked
    greenkeeper/prettier-1.10.2                  tracked
    greenkeeper/prettier-1.9.1                   tracked
    greenkeeper/prettier-1.9.2                   tracked
    greenkeeper/rollup-0.53.3                    tracked
    greenkeeper/rollup-0.53.4                    tracked
    greenkeeper/rollup-0.54.0                    tracked
    greenkeeper/rollup-0.54.1                    tracked
    greenkeeper/rollup-0.55.0                    tracked
    greenkeeper/rollup-0.55.1                    tracked
    greenkeeper/rollup-0.55.2                    tracked
    greenkeeper/rollup-0.55.3                    tracked
    greenkeeper/rollup-plugin-commonjs-8.3.0     tracked
    greenkeeper/rollup-plugin-node-resolve-3.0.2 tracked
    greenkeeper/selenium-assistant-5.3.0         tracked
    intial-pointer-fixup                         tracked
    master                                       tracked
    prettier                                     tracked
    readme-update                                tracked
    remove-dist                                  tracked
    ugh-eslint                                   tracked
  Local branch configured for 'git pull':
    master merges with remote master
  Local ref configured for 'git push':
    master pushes to master (up to date)
```

**Comment by mafredri** [comment]
@robdodson this exact error?

> prompt_pure_async_callback:79: bad math expression: illegal character: \

What git version are you using?

From the output of `GIT_TERMINAL_PROMPT=0 git remote show origin` you posted, it seems like there should be no problem (no username/password prompt).

**Comment by mafredri** [comment]
Would anyone who's able to reproduce this issue be willing to try the following patch?

```
diff --git a/pure.zsh b/pure.zsh
index 2fd19e8..82a6ce6 100644
--- a/pure.zsh
+++ b/pure.zsh
@@ -271,7 +271,7 @@ prompt_pure_async_git_fetch() {
 	# set ssh BachMode to disable all interactive ssh password prompting
 	export GIT_SSH_COMMAND=${GIT_SSH_COMMAND:-"ssh -o BatchMode=yes"}
 
-	command git -c gc.auto=0 fetch &>/dev/null || return 99
+	command git -c gc.auto=0 fetch | cat &>/dev/null || return 99
 
 	# check arrow status after a successful git fetch
 	prompt_pure_async_git_arrows $1
```

It runs `git fetch` through pipe-cat to make sure git does not try to spawn an interactive session. It should (hopefully) make the `bad math expression` error go away.

**Comment by mafredri** [comment]
Bummer :(, thanks for testing. I’m was hoping this would prevent interactive (e.g. ncurses-based) programs from starting in the async worker, since that’s not supported. That error came to be because the interactive program ate part of the output from the worker 😕.

I think I will have to try to reproduce this somehow to debug further. Any help in creating a reproduction setup would be much appreciated. (I have no experience of pinentry and the like.)

**Comment by davidtwco** [comment]
@mafredri this particular issue isn't related to ncurses (as best I can tell - that's more #366 which I've also run into). 

To reproduce this, create a VM (or something that you can ssh to) that requires password authentication, then set it as a remote for a repository. When you run commands in that folder, every so often (normally every second command or so), it outputs `prompt_pure_async_callback:79: bad math expression: illegal character: \`.

I ran into it when I was working on a project that required jumping through a bastion host that only allowed password authentication.

In my particular case, the ssh host is defined in `.ssh/config` and uses the `ProxyJump` directive - I doubt this makes a difference, but if the above doesn't repro then it might be worth trying.

**Comment by mafredri** [comment]
> this particular issue isn't related to ncurses (as best I can tell - that's more #366 which I've also run into).

I realize, bringing up ncurses was a mistake. Still, either SSH is interactive (prompting for password) or some other is.

I couldn't reproduce this, but it got me thinking that perhaps you have defined a custom `GIT_SSH_COMMAND`? Pure, perhaps foolishly, prioritizes the users choice here.

How about this patch that always forces batch mode?

```patch
diff --git a/pure.zsh b/pure.zsh
index 01c1980..9a4ec12 100644
--- a/pure.zsh
+++ b/pure.zsh
@@ -272,7 +272,7 @@ prompt_pure_async_git_fetch() {
 	# set GIT_TERMINAL_PROMPT=0 to disable auth prompting for git fetch (git 2.3+)
 	export GIT_TERMINAL_PROMPT=0
 	# set ssh BachMode to disable all interactive ssh password prompting
-	export GIT_SSH_COMMAND=${GIT_SSH_COMMAND:-"ssh -o BatchMode=yes"}
+	export GIT_SSH_COMMAND="${GIT_SSH_COMMAND} -o BatchMode=yes"
 
 	command git -c gc.auto=0 fetch &>/dev/null || return 99
 

```

**Comment by davidtwco** [comment]
@mafredri I believe it is SSH prompting for a password. I don't have a `GIT_SSH_COMMAND` defined. 

Started noticing that sometimes it says `prompt_pure_async_callback:79: bad math expression: operand expected at end of string` as well as `prompt_pure_async_callback:79: bad math expression: illegal character: \`.

However, I think that does seem to be fixing it (with the last patch too, I haven't removed that).

**Comment by mafredri** [comment]
@davidtwco I would appreciate it if you could give #392 a try, by itself (no other patches applied).

**Comment by davidtwco** [comment]
@mafredri sadly, this doesn't seem to be working, I checked out your branch and I keep seeing `prompt_pure_async_callback:79: bad math expression: operand expected at end of string` and `prompt_pure_async_callback:79: bad math expression: illegal character: \` every second command or so. 

I even tried re-adding the `cat` addition (since I had both before when it was working) and that doesn't seem to be doing it now - not sure what I was observing before.

**Comment by mafredri** [comment]
I had amended my patch from above, it was incorrect:

Was:
```
GIT_SSH_COMMAND="${GIT_SSH_COMMAND} -o BatchMode=yes"
```

Is now:
```
GIT_SSH_COMMAND="${GIT_SSH_COMMAND:-"ssh"} -o BatchMode=yes"
```

The reason that the original patch might have worked is that SSH was never launched (due to `GIT_SSH_COMMAND=" -o BatchMode=yes"` being an invalid command).

I'm pretty sure the `cat` patch from before did nothing to help the situation.

So the question is, how do we disable this for you? I do believe your `ProxyJump` is key here, it's invoking SSH without `-o BatchMode=yes`. The result being =>

```
ssh batch -> jump -> ssh (prompt password)
```

**Comment by davidtwco** [comment]
I'm not sure this should have been closed, if I'm not mistaken #392 attempted to fix but wasn't quite working. @mafredri

**Comment by davidtwco** [comment]
@binaryplease I believe #397 fixed this.
